pytest-embedded/pytest_embedded/dut_factory.py

```diff
@@ -0,0 +1,725 @@
+        if isinstance(obj, (subprocess.Popen, multiprocessing.process.BaseProcess)):
+            obj.terminate()
+            obj.kill()
+        elif isinstance(obj, io.IOBase):
+            try:
+                obj.close()
+            except Exception as e:
+                logging.debug('file %s closed failed with error: %s', obj, str(e))
+        else:
+            try:
+                obj.close()
+            except AttributeError:
+                try:
+                    obj.terminate()
+                except AttributeError:
+                    pass
+            except Exception as e:
+                logging.debug('Not properly caught object %s: %s', obj, str(e))
+    except Exception as e:
+        logging.debug('%s: %s', obj, str(e))
+        return  # swallow up all error
+    finally:
+        referrers = gc.get_referrers(obj)
+        for _referrer in referrers:
+            if isinstance(_referrer, list):
+                for _i, val in enumerate(_referrer):
+                    if val is obj:
+                        _referrer[_i] = None
+            elif isinstance(_referrer, dict):
+                for key, value in _referrer.items():
+                    if value is obj:
+                        _referrer[key] = None
+        del obj
+
+
+class DutFactory:
+    obj_stack: t.ClassVar[t.List[t.List[t.Any]]] = []
+
+    @classmethod
+    def close(cls):
+        global dut_global_index
+        dut_global_index = 0
+        if hasattr(cls, 'obj_stack'):
+            while cls.obj_stack:
+                layout = DutFactory.obj_stack.pop()
+                while layout:
+                    obj = layout.pop()
+                    _close_or_terminate(obj)
+                del layout
+            del DutFactory.obj_stack
+        cls.obj_stack = []
+
+    @classmethod
+    def unity_tester(cls, *args: 'IdfDut'):
+        return CaseTester(args)
+
+    @classmethod
+    def create(
+        cls,
+        *,
+        embedded_services='',
+        app_path='',
+        build_dir='build',
+        port=None,
+        port_location=None,
+        port_mac=None,
+        target=None,
+        beta_target=None,
+        baud=None,
+        skip_autoflash=None,
+        erase_all=None,
+        esptool_baud=None,
+        esp_flash_force=False,
+        part_tool=None,
+        confirm_target_elf_sha256=None,
+        erase_nvs=None,
+        skip_check_coredump=None,
+        panic_output_decode_script=None,
+        openocd_prog_path=None,
+        openocd_cli_args=None,
+        gdb_prog_path=None,
+        gdb_cli_args=None,
+        no_gdb=None,
+        qemu_image_path=None,
+        qemu_prog_path=None,
+        qemu_cli_args=None,
+        qemu_extra_args=None,
+        wokwi_cli_path=None,
+        wokwi_timeout=0,
+        skip_regenerate_image=None,
+        encrypt=None,
+        keyfile=None,
+    ):
+        """
+        Create a Device Under Test (DUT) object with customizable parameters.
+
+        Parameters:
```

Comment:
Here you're writing documentation in a non-standard way. This repo is using google docstring style. https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#:~:text=def%20function_with_pep484_type_annotations(,False%20otherwise.%0A%0A%20%20%20%20%22%22%22

besides, it's much better to write the types in the function signatures than in the docstrings.